draft-ietf-ppm-dap.md

```diff
@@ -852,14 +867,21 @@ provide for some small leeway, usually no more than a few minutes, to account
 for clock skew. If the leader rejects a report for this reason, it SHOULD abort
      the upload protocol and alert the client with error "reportTooEarly".
 
+If the Report contains an unrecognized exgension, then the Leader MAY abort the
+upload request with error "unrecognizedExtension". Note that this behavior is
+not mandatory because it requires the Leader to decrypt its input share.
+
 ### Upload Extensions {#upload-extensions}
 
-Each Report carries a list of extensions that clients may use to convey
-additional, authenticated information in the report. [OPEN ISSUE: The extensions
-aren't authenticated. It's probably a good idea to be a bit more clear about how
-we envision extensions being used. Right now this includes client attestation
-for defeating Sybil attacks. See issue#89.] Each extension is a tag-length
-     encoded value of the following form:
+Each Report carries lists of extensions that Clients use to convey additional,
+authenticated information in ts Report. Extensions that need to be handled by
+all Aggregators are included in the ReportMetadata; extensions intended to be
+handled by a specific Aggregator are included in the PlaintextInputShare. (For
```

Comment:
I worry that putting extensions in two different places will lead to complexity in the implementations and ambiguities, too: what if the same extension type appears in both places but with different values? Either DAP or the specification of each extension has to spell out how to resolve such conflicts.

Unless I'm missing something, the reason to allow extensions in two places is to avoid transmitting the same extension in each aggregator's ciphertext. If I'm right and this is just about that performance optimization, could we avoid doing this until operational experience shows it to be a problem worth the extra complexity?

Reply:
We already have this problem for ReportMetadata.extensions, but I agree that de-duplicating in two places adds a bit more complexity. Basically instead of:

1. Parse report
2. De-duplicate Report.metadata.extensions
3. Decrypt input share

We have to do

1. Parse report
2. Decrypt input share
3. De-duplicate Report.metadata.extensions and plaintext_input_share.extensions

> Unless I'm missing something, the reason to allow extensions in two places is to avoid transmitting the same extension in each aggregator's ciphertext. If I'm right and this is just about that performance optimization, could we avoid doing this until operational experience shows it to be a problem worth the extra complexity?

Given how important a consideration bandwidth has been, I think it's worth adding a bit of complexity to avoid increasing overhead here. This is particularly concerning for the taskprov extension, FWIW.

Reply:
I'm still concerned about this, but I'm not sure how to break the tie between us. I'm OK moving forward with this change as it is (especially since you helpfully filed #375), but perhaps we could take this question to the WG to get more opinions?

Reply:
I think separate places for per-report and per-share extensions makes sense. We may want to add guidance in the extensions section or security considerations section for extension designers. Specifically, I think each extension should have to define whether it is allowable as a per-report extension, per-share extension, or both. If an extension implicitly relies on both aggregators seeing the same extension value, then failure to reject that extension inside `PlaintextInputShare` would be an easy implementation mistake to make.

Reply:
One datapoint from IETF 115: There is not much enthusiasm for taskprov extension as-it-is, so I think it would be fair to disregard that particular use case as a consideration. That said I imagine there might be use cases for which shared extensions are useful.

Let's take this to the list, as @tgeoghegan suggests. Hopefully someone will give us a perspective that makes the decision tree more clear.

Reply:
Update the PR to remove shared extensions.